File: python3/count_primes.py
# Space   : O(n)
# Time    : O(n log n)


class Solution:
    def countPrimes(self, n: int) -> int:
        if n <= 1:
            return 0
        
        isPrime = [True] * n
        isPrime[0] = False
        isPrime[1] = False

        i = 2
        while i*i < n:
            if not isPrime:
                continue
            j = i*i
            while j < n:
                isPrime[j] = False
                j += i
            i += 1

        count = 0
        for x in isPrime:
            if x:
                count += 1
            
        return count

File: golang/count_primes.go
package main

// Space   : O(n)
// Time    : O(n log n)

func countPrimes(n int) int {
	isPrime := []bool{}

	for a := 0; a < n; a++ {
		if a <= 1 {
			isPrime = append(isPrime, false)
		} else {
			isPrime = append(isPrime, true)
		}
	}

	for i := 0; i*i < n; i++ {
		if isPrime[i] == false {
			continue
		}
		for j := i * i; j < n; j += i {
			isPrime[j] = false
		}
	}

	count := 0
	for i := 0; i < n; i++ {
		if isPrime[i] == true {
			count++
		}
	}

	return count
}
